src/sentry/sentry_metrics/consumers/indexer/parallel.py

```diff
@@ -1,111 +1,180 @@
-        self.__max_batch_size = max_batch_size
+
+        # This is the size of the initial message batching the indexer does
+        self.__max_msg_batch_size = max_msg_batch_size
+        self.__max_msg_batch_time = max_msg_batch_time
+
+        # This is the size of the batches sent to the parallel processes.
+        # These are batches of batches.
+        self.__max_parallel_batch_size = max_parallel_batch_size
+        self.__max_parallel_batch_time = max_parallel_batch_time
+
+        # This is the batch size used to decide when to commit on Kafka.
+        self.__commit_max_batch_size = max_batch_size
+        self.__commit_max_batch_time = max_batch_time
 
         self.__processes = processes
 
         self.__input_block_size = input_block_size
         self.__output_block_size = output_block_size
 
-        self.__commit_max_batch_size = commit_max_batch_size
-        self.__commit_max_batch_time = commit_max_batch_time
-
     def create_with_partitions(
         self,
         commit: Callable[[Mapping[Partition, Position]], None],
         partitions: Mapping[Partition, int],
     ) -> ProcessingStrategy[KafkaPayload]:
         parallel_strategy = ParallelTransformStep(
             partial(process_messages, self.__config.use_case_id),
-            SimpleProduceStep(
-                commit_function=commit,
-                commit_max_batch_size=self.__commit_max_batch_size,
-                # convert to seconds
-                commit_max_batch_time=self.__commit_max_batch_time / 1000,
-                output_topic=self.__config.output_topic,
+            Unbatcher(
+                SimpleProduceStep(
+                    commit_function=commit,
+                    commit_max_batch_size=self.__commit_max_batch_size,
+                    # This is in seconds
+                    commit_max_batch_time=self.__commit_max_batch_time / 1000,
+                    output_topic=self.__config.output_topic,
+                ),
             ),
             self.__processes,
-            max_batch_size=self.__max_batch_size,
-            max_batch_time=self.__max_batch_time,
+            max_batch_size=self.__max_parallel_batch_size,
+            # This is in seconds
+            max_batch_time=self.__max_parallel_batch_time / 1000,
             input_block_size=self.__input_block_size,
             output_block_size=self.__output_block_size,
-            initializer=initializer,
+            # It is absolutely crucial that we pass a function reference here
+            # where the function lives in a module that does not depend on
+            # Django settings. `sentry.runner` fulfills that requirement, but
+            # if you were to create a wrapper function in this module that
+            # calls configure(), and pass that function here, it would attempt
+            # to pull in a bunch of modules that try to read django settings at
+            # import time
+            initializer=configure,
```

Comment:
damn it. so the issue was in unpickling the module for the initializer ?

Reply:
"yes", but at the risk of being extremely pedantic, i think there's a misconception on how functions are pickled. so i need to hold a lecture now, sorry.

you're saying the module is pickled, but if you pickle a function, no module is ever pickled. the function is effectively serialized as a tuple of `(myfn.__module__, myfn.__name__)` over the wire, and `pickle.loads` attempts to import that function normally.

this has two practical implications:

* global functions are "basically always" pickle-safe for example, it is fine to do this at the module level:

```
# a bunch of objects that cannot be pickled
redis_client = Client()
randomfile = open("/tmp/foo")

class PickleBreaker:
    def __setstate__(self, state):
        raise RuntimeError("NOPE")

foobar = PickleBreaker()

def myfn(): pass

pickle.dumps(myfn)  # works
```

each process will import their own module, and as a result have their own redis client and file descriptor.

* it makes closures un-pickleable generally, because those aren't "singletons" i.e. don't have a stable import path:

```
def get_myfn():
    def closure(): pass
    return pickle.dumps(closure)  # fails
```

----

what happened here is not that the global function was unsafe to pickle. it was just that we tried to import the sentry app without initializing it. we sent `("sentry.sentry_metrics.consumers.indexer.parallel", "initializer")` to the child process to import and call. you can open a bare python shell and reproduce this without pickle:

```
$ python
>>> from sentry.sentry_metrics.consumers.indexer.parallel import initializer # broken
>>> from sentry.runner import configure  # works
```